### ISSUE TITLE
ControllerMap doesn't work with subdirectory controllers

### ISSUE THREAD

**Issue by nganhtuan63** [yiisoft/yii2#7899]
Currently, I have a module 'base' where I have arranged my controllers under the 'admin' or 'site' folders. Below is the sample structure:

---yii2base (module)
---controllers
------admin
----------UserController.php
----------TenantController.php

I want to override the TenantController in my advanced app structure. And I have registered the controllerMap like this: 

![screen shot 2015-03-28 at 9 09 15](https://cloud.githubusercontent.com/assets/1411732/6879500/27a31242-d52a-11e4-84e4-aa82861937e0.png)

However, it doesn't work. I did some debugs and I have found that the yii/base/Module.php only check controllerMap[$id] which only return 'admin' part of the route while it should also check 'admin/tenant' route exists in controllerMap. 

What do you think?

**Comment by mdmunir** [comment]
This issue is valid. My suggestion

this code https://github.com/yiisoft/yii2/blob/master/framework/base/Module.php#L499-L504
```php
        if (strpos($route, '/') !== false) {
            list ($id, $route) = explode('/', $route, 2);
        } else {
            $id = $route;
            $route = '';
        }
```
can be changed with
```php
        $id = $route;
        $route = '';
        while (($pos = strrpos($id, '/')) !== false) {
            if (isset($this->controllerMap[$id])) {
                $controller = Yii::createObject($this->controllerMap[$id], [$id, $this]);
                return [$controller, $route];
            }
            if ($route === '') {
                $route = substr($id, $pos + 1);
            } else {
                $route = substr($id, $pos + 1).'/'.$route;
            }
            $id = substr($id, 0, $pos);
        }
```

**Comment by lynicidn** [comment]
show url rules

**Comment by mdmunir** [comment]
^ its not about url rule.
@samdark IMO its also not a bug. Its must be features request.
FR: `controllerMap` support for subdirectory.

**Comment by lynicidn** [comment]
@mdmunir:
1. url rule must break it
2. need revert slash

**Comment by mdmunir** [comment]
@lynicidn 
1. I am not sure what you mean. But, its not about url rule.
2. Ya, i know. @nganhtuan63 wrong when write controller id using backslash. But revert slash also not suported.

**Comment by lynicidn** [comment]
@mdmunir 
1. i mean <m>/<c>/<a>
3. also in ur code
```php
if ($route === '') {
```
$route allways be ''

**Comment by mdmunir** [comment]
1. Just let `url rule` as empty (use default setting). 
2. Thats why i say, its enhancment or FR. I want controlerMap support for subdirectory
e.g
```php
'controllerMap' => [
    'user/register' => [
        'class' => 'app\controllers\user\RegisterController',
        'userModel' => 'my\ext\models\User',
    ]
]
```

**Comment by klimov-paul** [comment]
I am unsure about this issue.
`Module::controllerMap` serves ability to share controller defined elsewhere with the module without necessity of copying it.
Although `controllerMap` can be used for overriding existing controllers, this issue sounds like abuse of the original idea.

**Comment by mdmunir** [comment]
@klimov-paul 
Ignore my comments above. 
I just want to ask, Do you have plan to support subdirectory in `controllerMap`?

**Comment by klimov-paul** [comment]
That is not for me to decide. At least, not for me alone.

**Comment by samdark** [comment]
I think it's good to have supported if it won't result in singificant performance drop.

**Comment by qiangxue** [comment]
This is not trivial. Imagine you have a URL `a/b/c/d`. When in module `a`, you need to test `b`, `b/c`, `b/c/d`. Note that this lies on the critical path.

**Comment by mdmunir** [comment]
@qiangxue 
For now, we suport subdirectory both for application and module. will we suport for `controllerMap` to?
If you ask, do you using subdirectory for module. I said yes, i do.

**Comment by petrabarus** [comment]
Vote up for this issue. :+1:

A lot of times I like to ship `Controller` class as simple extension instead of big module. 
Currently my website uses twitter like username directory, e.g. `www.example.com/petrabarus`.
Without supporting a slash in the `controllerMap` name, I have to add the controller name to the `username` blacklist to prevent user registering as the user. If the `slash` is supported, I just need to define one `username` blacklist and attach the `Controller` as the subdirectory.

But still, this is trivial issue. Might like to see this gets implemented if not dropping performance.

**Comment by samdark** [comment]
Won't be implemented in 2.0.
